plugins/wasmembed/secure-enclave.cpp

```diff
@@ -26,97 +24,137 @@
+    void setCodeContext(ICodeContext *codeCtx)
     {
-        std::shared_lock lock(mutex);
-        auto it = map.find(key);
-        if (it != map.end())
-        {
-            value = it->second;
-            return true;
-        }
-        return false;
+        TRACE("WASM SE setCodeContext");
+        std::call_once(wasmLoadedFlag, &WasmEngine::loadWasmFiles, this, codeCtx);
     }
 
-    bool has(const K &key) const
+    bool hasModule(const std::string &wasmName) const
     {
-        std::shared_lock lock(mutex);
-        return map.find(key) != map.end();
+        TRACE("WASM SE hasModule");
+        return wasmModules.find(wasmName) != wasmModules.end();
     }
 
-    void forEach(std::function<void(const K &key, const V &value)> func) const
+    wasmtime::Module getModule(const std::string &wasmName) const
     {
-        std::shared_lock lock(mutex);
-        for (auto it = map.begin(); it != map.end(); ++it)
-        {
-            func(it->first, it->second);
-        }
+        TRACE("WASM SE getModule");
+        auto found = wasmModules.find(wasmName);
+        if (found == wasmModules.end())
+            throw makeStringExceptionV(1, "Wasm module not found: %s", wasmName.c_str());
+        return found->second;
     }
 };
+static std::unique_ptr<WasmEngine> wasmEngine = std::make_unique<WasmEngine>();
 
-class WasmEngine
+class WasmStore
 {
 private:
-    wasmtime::Engine engine;
     wasmtime::Store store;
 
-    ThreadSafeMap<std::string, wasmtime::Instance> wasmInstances;
-
-    //  wasmMems and wasmFuncs are only written to during createInstance, so no need for a mutex.
-    //  createInstance is only called from within wasmInstances.insertIfMissing via lambda function
-    //  So is behind a mutex anyway.
+    std::unordered_map<std::string, wasmtime::Instance> wasmInstances;
     std::unordered_map<std::string, wasmtime::Memory> wasmMems;
     std::unordered_map<std::string, wasmtime::Func> wasmFuncs;
 
-private:
-    //  Do not call this function directly...
-    wasmtime::Instance createInstance(const std::string &wasmName, const std::variant<std::string_view, wasmtime::Span<uint8_t>> &wasm)
+public:
+    WasmStore() : store(wasmEngine->engine)
+    {
+        TRACE("WASM SE WasmStore");
+    }
+
+    ~WasmStore()
+    {
+        TRACE("WASM SE ~WasmStore");
+    }
+
+    bool hasInstance(const std::string &wasmName) const
     {
+        TRACE("WASM SE hasInstance");
+        return wasmInstances.find(wasmName) != wasmInstances.end();
+    }
+
+    wasmtime::Instance getInstance(const std::string &wasmName) const
+    {
+        TRACE("WASM SE getInstance");
+        auto found = wasmInstances.find(wasmName);
+        if (found == wasmInstances.end())
+            throw makeStringExceptionV(1, "Wasm instance not found: %s", wasmName.c_str());
+        return found->second;
+    }
+
+    void registerInstance(const std::string &wasmName)
+    {
+        TRACE("WASM SE registerInstance %s", wasmName.c_str());
```

Comment:
Are there thread safety issues here?  Currently thread local, so no, but will that cause other issues?

Reply:
I don't believe so, this PR reworked the engine / store threading specifically to match the threaded examples in the wasmtime repo (engine is a global and store / store context is per thread).